circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,158 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+
+use crate::{params::RoundParams, poseidon::native::sponge::PoseidonSponge};
+
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	_params: PhantomData<P>,
+	root: Option<F>,
+	indexes: Vec<Option<F>>,
+	height: usize,
+	first_leaf_index: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	fn new(height: usize) -> Self {
+		let mut first_leaf_index = 1;
+		let mut nodes = 1;
+
+		for i in 0..height {
+			first_leaf_index *= 2;
+			nodes = 2 * nodes;
+		}
+		first_leaf_index -= 1;
+		nodes *= 2;
+		let indexes = vec![None; nodes - 1];
+		MerkleTree { _params: PhantomData, root: None, indexes, height, first_leaf_index }
+	}
+
+	fn implement_leaves(self: &mut MerkleTree<F, P>, values: Vec<F>) -> &Self {
+		for i in 0..values.len() {
+			self.indexes[self.first_leaf_index + i] = Some(values[i]);
+		}
+		self
+	}
+
+	fn build_tree(self: &mut MerkleTree<F, P>, height: usize) -> &Self {
+		if height > 0 {
+			let mut new_tree = MerkleTree::<F, P>::new(height - 1);
+			let first_leaf_index = (new_tree.first_leaf_index * 2) + 1;
+			let mut hash = Vec::new();
+			for i in 0..first_leaf_index {
+				let mut hasher: PoseidonSponge<F, WIDTH, P> = PoseidonSponge::new();
```

Comment:
Not sure why we are using sponge, when we are hashing only 2 values.

Reply:
I did use this because Poseidon accepts minimum width of 5. I thought I can construct an array with width 5 but we will only use 2 for each loop. Is this better use case then the sponge?

Reply:
I think we should use sponge for arbitrary length arrays where length is > WIDTH. In cases like this, when we are sure about our input width, we could use normal poseidon and add padding manually, e.g.:
`[value1, value2, zero, zero, zero]`